### ISSUE TITLE
Get Mali working on X11 and Android

### ISSUE THREAD

**Issue by zenitraM** [linux-sunxi/linux-sunxi#19]
Mali DRM required module has been integrated and is now on the wip/allwinner-v3.0-android-v2/mali branch.

However, using the xserver-xorg-mali package from Linaro PPA:

```

X.Org X Server 1.11.3
Release Date: 2011-12-16
X Protocol Version 11, Revision 0
Build Operating System: Linux 2.6.38-1209-omap4 armv7l Linaro
Current Operating System: Linux linaro-ubuntu-desktop 3.0.31+ #6 PREEMPT Sat May 19 01:36:40 CEST 2012 armv7l
Kernel command line: console=ttyS0,115200 root=/dev/mmcblk0p2 init=/init loglevel=8 rootfstype=ext4 rootwait
Build Date: 25 April 2012  06:26:32PM
xorg-server 2:1.11.4-0ubuntu10.1linaro1 (For technical support please see http://www.ubuntu.com/support) 
Current version of pixman: 0.24.4
    Before reporting problems, check http://wiki.x.org
    to make sure that you have the latest version.
Markers: (--) probed, (**) from config file, (==) default setting,
    (++) from command line, (!!) notice, (II) informational,
    (WW) warning, (EE) error, (NI) not implemented, (??) unknown.
(==) Log file: "/var/log/Xorg.0.log", Time: Fri May 18 23:44:04 2012
(==) Using system config directory "/usr/share/X11/xorg.conf.d"
mali trace: MaliSetup
mali trace: MaliIdentify
mali trace: MaliProbe
mali trace: MaliHWProbe
mali trace: MaliPreInit
mali trace: MaliGetRec
mali trace: MaliHWInit
mali trace: MaliHWGetDepth
mali trace: HWGetVidmem
mali trace: MaliHWGetName
mali trace: MaliHWGetFD
mali trace: FBDev_crtc_config
mali: FBDev_crtc_config done
mali trace: MaliScreenInit
(EE) MALI(0): mali_drm_open_master DRM OPEN (fd: 0x9)
mali trace: MaliHWMapVidmem
mali trace: MaliHWLinearOffset
mali trace: MaliHWSave
mali trace: MaliHWModeInit
mali trace: MaliHWSetMode
mali trace: MaliHWSaveScreen
mali trace: MaliHWAdjustFrame
libblt_hwCould not open device /dev/b2r2_blt
libblt_hwCould not open device /dev/b2r2_blt
(EE) MALI(0): XVScreenInit failed

Fatal server error:
AddScreen/ScreenInit failed for driver 0


Please consult the The X.Org Foundation support 
     at http://wiki.x.org
 for help. 
Please also check the log file at "/var/log/Xorg.0.log" for additional information.

mali trace: MaliHWLeaveVT
mali trace: MaliHWRestore
 ddxSigGiveUp: Closing log
Server terminated with error (1). Closing log file.
root@linaro-ubuntu-desktop:/usr/share/X11/xorg.conf.d# Write failed: Broken pipe
```

TODO: Check with the Mali Xorg sources from ARM to see if they also require this driver (i'm not sure if it's a HW component part of the Mali), and if so, it can be implemented from the GPL sources on Linaro tree: http://www.igloocommunity.org/gitweb/?p=kernel/igloo-kernel.git;a=tree;f=drivers/video/b2r2;h=424976ba6cce1abfa0f96048475a208b329f24d6;hb=66abcc2e034602ba5a0743475c41996c903caedd

**Comment by ghost** [comment]
b2r2 is the 2D accelerator on the Nova SoC on the Snowball board. The Mali xorg driver from ARM has stubs that must be populated to integrate with whatever display engine's present on the target.

**Comment by j1nx** [comment]
I did the exact same thing the other day, but did not finish it, nor test it.

Some other things todo, which might help.
1) Upgrade the mali.ko and ump.ko with the exact same version as the mali_drm.ko, by downloading http://www.malideveloper.com/files/DX910-SW-99002-r2p4-02rel1.tgz
Put the files in "modules/mali/" and update the Makefile to point top the new folder. (you also have to copy the config file from the old version into the new version). This will compile a newer mali.ko and ump.ko

2) In the rootfs from Tom. In the "/home/ubuntu/mali" directory their are all the closed sourced libraries for the same version (r2p4-02rel1).

3) Next will then be to download and compile the X11 driver from malidevelopers. (Also avaliable into the same folder as above)

**Comment by hipboi** [comment]
EXA is an acceleration architecture that provides hooks for accelerating 2D graphics operations
in the X display driver.

The X.Org installation typically includes a library called Pixman that includes software
optimized versions of these operations. If the EXA acceleration hooks are not implemented the
XServer uses Pixman for 2D graphics operations.

The 2d acceleration is optional

**Comment by amery** [comment]
I've merged into this branch the DX910-SW-99002-r2p4-02rel0 device drivers from @Turl, he is using this version succesfully on android. I've also merged the HEAD of 3.0-v2 into the branch to collect the fixes so people can actually use this branch

**Comment by j1nx** [comment]
Ok nice, but ....

@Turl; What made you choose to upgrade to an old version? If you upgrade, why not do it to the latest version available from malidevelopers. (r2p4-02rel1)
http://www.malideveloper.com/developer-resources/drivers/open-source-mali-gpus-linux-kernel-device-drivers.php

**Comment by amery** [comment]
@j1nx, I was told by @Turl on IRC it was because that was the version of the userspace libraries he had

**Comment by j1nx** [comment]
Tom released the r2p4-02rel1 userspace libraries in his Ubuntu image. (In the ubuntu user his home directory).

Tom also is working on the X11 drivers and I asked him nicely to utilise the latest and greatest version from malidevelopers.com

He stated he will. So we are better of updating to the latest version, maybe put the userspace libraries from the ubuntu release into a seperate "lib" folder here on github. (Tom made a statement by email we are allowed to use them. That statement was also copied to the rhombus-tech wiki I thought).

Then when Tom compiles and release the X11 driver from the same version using the Mali DDK he has on his box, then we should be fine.

We can then even copy over the DRM drivers from the same r2p4-02rel1 release and appling the drm patch. Than we got is all at the latest version. (Maybe set the DRM settings to "no" by default.)

**Comment by amery** [comment]
a complete upgrade to r2p4-02rel1 from @Turl has been merged, hopefully we can test linux support soon. feedback appreciated

**Comment by TJC** [comment]
Hi,
I have successfully used the Mali device driver on Linux, using the android-3.0-v2 branch here.

I followed the instructions at http://rhombus-tech.net/allwinner_a10/Compile_X11_driver_for_A10/

**Comment by TJC** [comment]
```
[    23.899] (WW) Falling back to old probe method for mali
[    23.914] MaliProbe
[    23.914] MaliHWProbe
[    23.914] (II) Loading /usr/lib/xorg/modules/drivers/mali_drv.so
[    23.914] (II) MALI(0): using /dev/fb0
.....
[    23.915] (II) MALI(0): hardware:  (video memory: 16200kB)
.....
[    23.944] (II) MALI(0): Mali EXA driver is loaded successfully
[    23.944] (II) EXA(0): Driver allocated offscreen pixmaps
[    23.944] (II) EXA(0): Driver registered support for the following operations:
[    23.944] (II)         Solid
[    23.944] (II)         Copy
[    23.944] (II)         Composite (RENDER acceleration)
......
```

**Comment by hipboi** [comment]
On Thu, Jul 19, 2012 at 3:55 PM, Toby Corkindale
reply@reply.github.com
wrote:

> ```
> [    23.899] (WW) Falling back to old probe method for mali
> [    23.914] MaliProbe
> [    23.914] MaliHWProbe
> [    23.914] (II) Loading /usr/lib/xorg/modules/drivers/mali_drv.so
> [    23.914] (II) MALI(0): using /dev/fb0
> .....
> [    23.915] (II) MALI(0): hardware:  (video memory: 16200kB)
> .....
> [    23.944] (II) MALI(0): Mali EXA driver is loaded successfully
> [    23.944] (II) EXA(0): Driver allocated offscreen pixmaps
> [    23.944] (II) EXA(0): Driver registered support for the following operations:
> [    23.944] (II)         Solid
> [    23.944] (II)         Copy
> [    23.944] (II)         Composite (RENDER acceleration)
> EXA is the 2D acceleration interface, which can be configured as
> software or hardware, default software on A10. So the log doesn't mean
> the 3D acceleration.
> ......
> ```
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/amery/linux-allwinner/issues/19#issuecomment-7091927

## 

Keep simple, stay foolish.

**Comment by TJC** [comment]
I know.. but at least the driver is loading and working in 2D, right?

**Comment by hipboi** [comment]
On Thu, Jul 19, 2012 at 4:41 PM, Toby Corkindale
reply@reply.github.com
wrote:

> I know.. but at least the driver is loading and working in 2D, right?
> yes, software 2D.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/amery/linux-allwinner/issues/19#issuecomment-7092738

## 

Keep simple, stay foolish.

**Comment by TJC** [comment]
What's the next step then?

**Comment by j1nx** [comment]
Basically: Waiting untill the A2000 box arrives at the ARM office and the ARM people had a good look at the issue.

**Comment by amery** [comment]
and ARM people is going to help _us_ or Allwinner's private tree?

I just gave up in changing the hack they do to "reserve" the memory for mali... it seems we'll have to use @Turl's hack like it or not.

**Comment by j1nx** [comment]
That depends on if the error lies in either the; Mali kernel driver, the UMP kernel driver, the xf86 mali driver or the closed libraries.

**Comment by danielg4** [comment]
The mali_drm module never initializes…

**Comment by TJC** [comment]
@danielg4 - I'm not an expert, but drivers/gpu/drm/mali/mali_drv.c doesn't seem to contain any code anyway.

**Comment by rzk** [comment]
@TJC DRM is fixed in c8c9ae7ef2bec34dc529bba763f36e8d3b1b4266 also I've wrote simple tutorial for compiling mali from sources, now it works with DRI2 with default configuration: http://linux-sunxi.org/Mali400#Mali-400_X11_DRI2_drivers

Testers and people who know linux graphics stack are strongly welcome.

**Comment by techn** [comment]
I think this can be closed. Mali drivers are loaded by X and 3D is working. For 2D acceleration we need new ticket.
http://linux-sunxi.org/Mali400
